lib/factory_bot/definition.rb

```diff
@@ -111,6 +111,15 @@ def callback(*names, &block)
 
     def base_traits
       @base_traits.map { |name| trait_by_name(name) }
+    rescue KeyError => error
+      raise append_definition_name_to_error(error)
+    end
+
+    def append_definition_name_to_error(error)
+      message = "#{error.message} (referenced within \"#{name}\" definition)"
+      error.class.new(message).tap do |new_error|
```

Comment:
This is a clever and effective way to re-raise the underlying `KeyError`.

Is there a downside to raising an error class that FactoryBot declares, instead of `KeyError` from Ruby's standard library?

Reply:
Sean beat me to this. I don't think it's a bad idea necessarily to use a `KeyError` but a FactoryBot error class would be sweet considering this seems like a FactoryBot specific error

Reply:
Do we consider the originally raised `KeyError` part of a factory's public interface? Is re-raising the `KeyError` motivated by backwards compatibility?

Reply:
The reason we are using a `KeyError` rather than a custom FactoryBot error class is because `KeyError`s come with `Did you mean?` suggestions (see b86ec13).

But you are reminding me that I should probably add a test around that case. The `Did you mean?` message is meant to come at the end, so I might have to insert the custom bit here into the middle somewhere.

Reply:
OK, I modified this to handle the `Did you mean?` message.

Reply:
As I think about this more, I think we could probably update the `FactoryBot::Registry` to raise a custom factory_bot error instead of a `KeyError`, since by the time we [build the new error](https://github.com/thoughtbot/factory_bot/blob/98790602896bbe93195cface4b448d859dae3721/lib/factory_bot/registry.rb#L42) the `Did you mean?` message is already populated.

Reply:
Sure I'll take a stab at that this weekend.